### PR TITLE
create class SAMRAIDataCache to cache allocated Eulerian data

### PR DESCRIPTION
**Pull request by boyceg**
This is a draft implementation of a SAMRAI data caching class (see #533).

### CODE REVIEW

**Comment by drwells** [comment]
:+1:

I will take a look tomorrow or Friday (once I am back from VT)

File: ibtk/src/utilities/SAMRAIDataCache.cpp [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    // TODO: Make this more generic and extensible.
+    bool convertable = false;
+    if (can_convert_to<CellVariable<NDIM, T> >(var))
+    {
+        Pointer<CellDataFactory<NDIM, T> > pdat_fac = patch_descriptor->getPatchDataFactory(idx);
+        TBOX_ASSERT(pdat_fac);
+        depth = pdat_fac->getDefaultDepth();
+        ghost_width = pdat_fac->getGhostCellWidth().max();
+        TBOX_ASSERT(ghost_width == pdat_fac->getGhostCellWidth().min());
+        convertable = true;
```
Comment (drwells) [inline ibtk/src/utilities/SAMRAIDataCache.cpp]: I think that we could refactor this internal block (which is common to every case in the `if/else if` chain) to something like
```cpp
template <typename FactoryType>
std::pair<int, int> // depth, ghost_width
get_characteristics(const PatchDescriptor<NDIM> &patch_descriptor,
                    const int idx)
{
    auto Pointer<FactoryType> = patch_descriptor->getPatchDataFactory(idx);
    depth = pdat_fac->getDefaultDepth();
    ghost_width = pdat_fac->getGhostCellWidth().max();
    TBOX_ASSERT(ghost_width == pdat_fac->getGhostCellWidth().min());
    return std::make_pair(depth, ghost_width);
}
```

**Comment by drwells** [comment]
This looks good. Could you add a simple test?

**Comment by boyceg** [comment]
Yes. I have run tests with #572 and everything seems to work as expected.

@drwells, if #572 looks good to you, can you run the tests on it? (Can we set up Jenkins to do this for us automatically?)

**Comment by drwells** [comment]
Is it good to go (aside from the testers) now?

**Pull request merged**